<h1 style='font-size:40px'> AWS CLI, SDK, IAM Roles & Policies</h1>

<h2 style='font-size:30px'> AWS EC2 Instance Metadata</h2>
<div> 
    <ul style='font-size:20px'> 
        <li> 
            A AWS fornece uma API com a qual podemos consultar metadados sobre nossas instâncias EC2.
        </li>
        <li>
            Fazendo um `curl` a ela, conseguimos resgatar dados de Security Groups ou endereços de IP da máquina. 
        </li>
    </ul>
</div>

<h2 style='font-size:30px'> AWS CLI Profiles</h2>
<div> 
    <ul style='font-size:20px'> 
        <li> 
            A CLI da AWS nos permite usar múltiplas contas distintas, além da default. 
        </li>
        <li>
            Para cadastrar uma nova conta, invoque `aws configure --profile &lt account-alias&gt`.
        </li>
    </ul>
</div>

<figure>
    <center style='font-size:20px;margin-top:20px'> 
            <img src='img/12_profiles_files.png'>
        <figcaption style='font-size:15px'>
            Arquivos de configuração AWS, após o cadastro do novo perfil.
        </figcaption>
    </center>
</figure>

<div> 
    <ul style='font-size:20px'> 
        <li> 
            Por fim, caso queira executar um comando com o novo perfil, use a flag `--profile &lt account-alias&gt`.
        </li>
    </ul>
</div>

<h2 style='font-size:30px'> AWS CLI with MFA</h2>
<div> 
    <ul style='font-size:20px'> 
        <li> 
            Caso deseje, conseguimos criar sessões efêmeras de nossas contas com credenciais temporárias.
        </li>
        <li>
            Para isso, vamos ter que vincular um dispositivo MFA ao nosso perfil no IAM. Ele receberá uma ARN.
        </li>
        <li>
            Uma vez feito isso, invocamos o comando `aws sts get-session-token --serial-number &lt mfa-device-arn&gt --token &lt token-mfa&gt --duration-seconds n`. Use o código que estiver aparecendo no seu dispositivo MFA para completar a chamada. 
        </li>
    </ul>
</div>

<figure>
    <center style='font-size:20px;margin-top:20px'> 
            <img src='img/12_cli_mfa.png'>
        <figcaption style='font-size:15px'>
            Output do comando, com as credenciais.
        </figcaption>
    </center>
</figure>

<h2 style='font-size:30px'> AWS SDK Overview</h2>
<div> 
    <ul style='font-size:20px'> 
        <li> 
            Apresentação geral da SDK da AWS.
        </li>
    </ul>
</div>

<h2 style='font-size:30px'> Exponential Backoff & Service Limit Increase</h2>
<div> 
    <ul style='font-size:20px'> 
        <li> 
            A AWS possui uma orientação de usar Exponential Backoff mediante erros de throttling ou do tipo 5xx.
        </li>
    </ul>
</div>

<h2 style='font-size:30px'> AWS Credentials Provider & Chain</h2>
<div> 
    <ul style='font-size:20px'> 
        <li> 
            Aula em que é mostrada a ordem de precedência do uso de credenciais.
        </li>
    </ul>
</div>

<figure>
    <center style='font-size:20px;margin-top:20px'> 
            <img src='img/12_credentials_hierarchy.png'>
    </center>
</figure>

In [2]:
! git add .
! git commit -am 'Pesquisar sobre o Sigv4 para fazer as anotações da Aula 135'
! git push

[master ff073ab] Aula 132
 4 files changed, 368 insertions(+), 2914 deletions(-)
 rewrite .ipynb_checkpoints/12_cli_sdk-checkpoint.ipynb (98%)
 rewrite 12_cli_sdk.ipynb (98%)
 create mode 100644 img/12_cli_mfa.png
 create mode 100644 img/12_profiles_files.png
Enumerating objects: 11, done.
Counting objects: 100% (11/11), done.
Delta compression using up to 24 threads
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 167.87 KiB | 20.98 MiB/s, done.
Total 7 (delta 4), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To https://github.com/felipesveiga/aws-developer.git
   6543750..ff073ab  master -> master


<p style='color:red'> Vi Aula 132 a 135; Pesquisar sobre o Sigv4 para fazer as anotações da Aula 135</p>